Helper to augment graph with additional deps

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163960
* __->__ #163959
* #163754
* #163215

In comm-compute overlap we will have a graph with: 

```
def foo(...):
     ag = all_gather(...)
     hiding_compute = mm(...)
     wait(ag)
```

There is no explicit dependency between the hiding compute and the collectives, but we want to add implicit dependencies from wait->hiding_compute, and from hiding_compute->all_gather to preserve overlap. 

Additionally, while bucketing, we will merge collective starts and collective waits together. In this case, we will want to treat the two nodes as a single subgraph - each node in the merged set will have the union of all deps in the set. 

This pr adds `AugmentedGraphHelper` that adds the apis, and allows querying for dependency with this augmented graph.




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

eellison committed:
Helper to augment graph with additional deps

[ghstack-poisoned]

eellison committed:
Update on "Helper to augment graph with additional deps"


In comm-compute overlap we will have a graph with: 

```
def foo(...):
     ag = all_gather(...)
     hiding_compute = mm(...)
     wait(ag)
```

There is no explicit dependency between the hiding compute and the collectives, but we want to add implicit dependencies from wait->hiding_compute, and from hiding_compute->all_gather to preserve overlap. 

Additionally, while bucketing, we will merge collective starts and collective waits together. In this case, we will want to treat the two nodes as a single subgraph - each node in the merged set will have the union of all deps in the set. 

This pr adds `AugmentedGraphHelper` that adds the apis, and allows querying for dependency with this augmented graph.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

eellison commented on test/inductor/test_augmented_graph_helper.py:
```diff
@@ -0,0 +1,341 @@
+
+    def get_deps(self, node):
+        """Helper to get dependencies for a node."""
+        return list(node.args) if hasattr(node, "args") else []
```
i'll blame claude here haha..

eellison commented on torch/_inductor/augmented_graph_helper.py:
```diff
@@ -0,0 +1,113 @@
+        for n2 in nodes[1:]:
+            self.merge_to_set(n1, n2)
+
+    def unmerge_node(self, node: fx.Node) -> None:
```
the use case is that to do a collective merge, we need to merge a node start to an existing set of starts, then, attempt to merge the waits. If the merging of waits is unsuccessful we undo the most recent adding of a node to a bucket.

I can assert that the new node is a singleton in the callee if that helps ? I agree that what you're suggesting might be more general but I don't want to complicate this with not yet real use cases yet.

v0i0 replied:
i see, had a look at the next pr in the stack. the rhs node in merge is always a singleton, so this is always fine there. also seems like `merge_nodes` is never called? maybe `merge_to_set` could assert `new_node` is singleton?

eellison replied:
sounds good

eellison committed:
Update on "Helper to augment graph with additional deps"


In comm-compute overlap we will have a graph with: 

```
def foo(...):
     ag = all_gather(...)
     hiding_compute = mm(...)
     wait(ag)
```

There is no explicit dependency between the hiding compute and the collectives, but we want to add implicit dependencies from wait->hiding_compute, and from hiding_compute->all_gather to preserve overlap. 

Additionally, while bucketing, we will merge collective starts and collective waits together. In this case, we will want to treat the two nodes as a single subgraph - each node in the merged set will have the union of all deps in the set. 

This pr adds `AugmentedGraphHelper` that adds the apis, and allows querying for dependency with this augmented graph.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

eellison committed:
Update on "Helper to augment graph with additional deps"


In comm-compute overlap we will have a graph with: 

```
def foo(...):
     ag = all_gather(...)
     hiding_compute = mm(...)
     wait(ag)
```

There is no explicit dependency between the hiding compute and the collectives, but we want to add implicit dependencies from wait->hiding_compute, and from hiding_compute->all_gather to preserve overlap. 

Additionally, while bucketing, we will merge collective starts and collective waits together. In this case, we will want to treat the two nodes as a single subgraph - each node in the merged set will have the union of all deps in the set. 

This pr adds `AugmentedGraphHelper` that adds the apis, and allows querying for dependency with this augmented graph.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

eellison committed:
Update on "Helper to augment graph with additional deps"


In comm-compute overlap we will have a graph with: 

```
def foo(...):
     ag = all_gather(...)
     hiding_compute = mm(...)
     wait(ag)
```

There is no explicit dependency between the hiding compute and the collectives, but we want to add implicit dependencies from wait->hiding_compute, and from hiding_compute->all_gather to preserve overlap. 

Additionally, while bucketing, we will merge collective starts and collective waits together. In this case, we will want to treat the two nodes as a single subgraph - each node in the merged set will have the union of all deps in the set. 

This pr adds `AugmentedGraphHelper` that adds the apis, and allows querying for dependency with this augmented graph.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

eellison committed:
Update on "Helper to augment graph with additional deps"


In comm-compute overlap we will have a graph with: 

```
def foo(...):
     ag = all_gather(...)
     hiding_compute = mm(...)
     wait(ag)
```

There is no explicit dependency between the hiding compute and the collectives, but we want to add implicit dependencies from wait->hiding_compute, and from hiding_compute->all_gather to preserve overlap. 

Additionally, while bucketing, we will merge collective starts and collective waits together. In this case, we will want to treat the two nodes as a single subgraph - each node in the merged set will have the union of all deps in the set. 

This pr adds `AugmentedGraphHelper` that adds the apis, and allows querying for dependency with this augmented graph.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]
